legends/load_legends.py

```diff
@@ -0,0 +1,309 @@
+import os
+import re
+from pathlib import Path
+from typing import Dict, List, Tuple
+
+import geopandas as gpd
+import numpy as np
+import pandas as pd
+import rasterio
+import requests
+import sqlalchemy as sqla
+from scipy.stats import qmc
+
+
+# retrieve colors in RGB format from the color table
+def get_colors(clrpath: Path) -> Dict[str, List[Tuple[float, float, float]]]:
+    SEP = ","
+    clrs = {}
+    with open(clrpath, mode="r") as clrfile:
+        # skip the header
+        _ = next(clrfile).split(SEP)
+        for line in clrfile:
+            key, *vals = line.split(SEP)
+            clrs[key] = [
+                tuple([int(v[2:]) / 255.0 for v in rgb.split()]) for rgb in vals
+            ]
+    return clrs
+
+
+# format the input string in order to be a valid filename
+def format_filename(input_str: str) -> str:
+    valid_fname = input_str.replace(" ", "")
+    valid_fname = re.sub(r"[^\w\s-]", "", valid_fname.lower())
+    valid_fname = re.sub(r"[-\s]+", "-", valid_fname).strip("-_")
+    return valid_fname
+
+
+# ------------------------------
+# settings and setup
+# ------------------------------
+
+# folders and files settings
+SRCDIR = Path(__file__).parent
+DATADIR = SRCDIR.parent / "data"
+BREAKSDIR = DATADIR / "breaks"
+CLRTABLE = "color-tables.csv"
+LAYERTABLE = "layers.csv"
+
+# breaks and layers settings
+ROW_LIMIT = 100000  # limit the data requests to the server
+MAX_LENGTH = 1000  # max height or width of a raster to be allowed to read the whole file in memory at once
+MAX_SIZE = 100000  # max size to sample a layer - to compute percentiles for legends
+SAMPLE_SIZE = (
+    512  # size of the sample of a single raster of the layer - use a multiple of 2
+)
+percentiles_list = [20, 40, 60, 80]  # percentiles used to compute legend breaks
+EXCLUDED = [
+    467,
+    474,
+]  # layers to exclude - custom legend has been set during data-integration
+API_KEY = os.environ.get(
+    "API_KEY"
+)  # private key required to upload the legends on the server
+
+# database settings
+DB_HOST = os.environ.get("DB_HOST")
+DB_PORT = os.environ.get("DB_PORT")
+DB_USER = os.environ.get("DB_USER")
+DB_PASSWORD = os.environ.get("DB_PASSWORD")
+DB_DB = os.environ.get("DB_DB")
+db_connection_url = (
+    "postgresql://{DB_USER}:{DB_PASSWORD}@{DB_HOST}:{DB_PORT}/{DB_DB}".format(
+        DB_HOST=DB_HOST,
+        DB_PORT=DB_PORT,
+        DB_USER=DB_USER,
+        DB_PASSWORD=DB_PASSWORD,
+        DB_DB=DB_DB,
+    )
+)
+
+if not os.path.exists(BREAKSDIR):
+    os.mkdir(BREAKSDIR)
+if not os.path.exists(LAYERTABLE):
+    raise FileNotFoundError(
+        "Layer list not found."
+    )
+if not os.path.exists(CLRTABLE):
+    raise FileNotFoundError(
+        "Color list not found."
+    )
+
+# read colors and layer names
+colors_dict = get_colors(CLRTABLE)
+layers = pd.read_csv(LAYERTABLE)
+layers["uploaded"] = 0
+
+# ------------------------------
+# computing breaks
+# ------------------------------
+
+for i, row in layers.iterrows():
```

Comment:
It would be useful to be able to add an argument (e.g., the "dataset_id"), if we do not want to calculate all legends but only replace the ones for the layers that are updated.

Reply:
Something like this:
```
if len(sys.argv) > 1:
    ds_ids = [int(x) for x in sys.argv[1:]]
    layers = layers.loc[layers["ds_id"].isin(ds_ids),:]
```